### ISSUE TITLE
 basic_string::_M_construct null not valid

### ISSUE THREAD

**Issue by devin122** [cmu-sei/pharos#197]
Running `ooanalyzer` crashes with `[INFO ]: Pharos main error: (std::logic_error) basic_string::_M_construct null not valid`.

Command line used was: `ooanalyzer --serialize=ooprog.ser --maximum-memory 128000 --prolog-facts=progfacts.pl --threads=-2 --per-function-timeout 60  ../app.exe`

Full error:
```
[INFO ]: Pharos main error: (std::logic_error) basic_string::_M_construct null not valid
OOAN[FATAL]: Backtrace:
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(+0x958d40) [0x7f0796dbcd40]
OOAN[FATAL]: | /lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa38c) [0x7f078fd2938c]
OOAN[FATAL]: | /lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa3f7) [0x7f078fd293f7]
OOAN[FATAL]: | /lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa6a9) [0x7f078fd296a9]
OOAN[FATAL]: | /lib/x86_64-linux-gnu/libstdc++.so.6(std::__throw_logic_error(char const*)+0x41) [0x7f078fd2021e]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(+0xb43e74) [0x7f0796fa7e74]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(pharos::SQLLiteApiDictionary::Data::get_db_function(sqlite3_stmt*) const+0xade) [0x7f0796fae86e]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(pharos::SQLLiteApiDictionary::Data::get_api_definition(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std:                                          :char_traits<char>, std::allocator<char> > const&) const+0x2d7) [0x7f0796faeff7]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(pharos::SQLLiteApiDictionary::get_api_definition(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const+0x26) [0x7f0796faf0c6]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(pharos::DLLStateCacheDict::get_api_definition(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const+0xb2) [0x7f0796faa6e2]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(std::vector<std::shared_ptr<pharos::APIDefinition const>, std::allocator<std::shared_ptr<pharos::APIDefinition const> > > pharos::ReportDictionary::get_definition<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const+0x75) [0x7f0796fbd695]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(pharos::ReportDictionary::get_api_definition(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const+0x2f) [0x7f0796fad99f]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(pharos::MultiApiDictionary::get_api_definition(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const+0x51) [0x7f0796fa8381]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(std::vector<std::shared_ptr<pharos::APIDefinition const>, std::allocator<std::shared_ptr<pharos::APIDefinition const> > > pharos::ReportDictionary::get_definition<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const+0x75) [0x7f0796fbd695]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(pharos::ReportDictionary::get_api_definition(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const+0x2f) [0x7f0796fad99f]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(pharos::OOAnalyzer::visit(pharos::FunctionDescriptor*)+0xae2) [0x7f07970dbbb2]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(+0xbaec95) [0x7f0797012c95]
OOAN[FATAL]: | /usr/local/bin/../lib/libpharos.so(+0xbb0339) [0x7f0797014339]
OOAN[FATAL]: | /lib/x86_64-linux-gnu/libboost_thread.so.1.71.0(+0x1143b) [0x7f078f6a943b]
Segmentation fault (core dumped)

```

**Comment by md5i** [comment]
I've pushed a commit that hopefully provides insight into this problem.  Based on the backtrace you supplied, this error is likely the result of trying to create a string from the result of `sqlite3_column_text()`.  Under normal circumstances this should not return `nullptr` unless the column contains a SQL `NULL`, but this is not necessarily the case if there is an out-of-memory condition in sqlite.  If this is what happened to you, this change should make that clear when it happens.

If this is not what is happening, I'd be interested to see what the result of running with this change is.  (I've employed much more defensive programming techniques around the call to `sqlite3_column_text()`.)

**Comment by sei-mwd** [comment]
I've pushed a commit that hopefully provides insight into this problem.  Based on the backtrace you supplied, this error is likely the result of trying to create a string from the result of `sqlite3_column_text()`.  Under normal circumstances this should not return `nullptr` unless the column contains a SQL `NULL`, but this is not necessarily the case if there is an out-of-memory condition in sqlite.  If this is what happened to you, this change should make that clear when it happens.

If this is not what is happening, I'd be interested to see what the result of running with this change is.  (I've employed much more defensive programming techniques around the call to `sqlite3_column_text()`.)
